### PR TITLE
feat: add rbac middleware system

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/service/http.go [hunk 1]
```diff
@@ -22,6 +23,7 @@ import (
 
 // Handler defines the HTTP handler used in CDS engine
 type Handler func(ctx context.Context, w http.ResponseWriter, r *http.Request) error
+type RbacChecker func(ctx context.Context, db *gorp.DbMap, vars map[string]string) error
```
Comment (fsamin) [inline engine/service/http.go:26]: *gorp.DbMap => SqlExecutor

File: engine/api/router_middleware_rbac.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	"net/http"
+
+	"github.com/gorilla/mux"
+
+	"github.com/ovh/cds/engine/service"
+)
+
+func (api *API) rbacMiddleware(ctx context.Context, _ http.ResponseWriter, req *http.Request, rc *service.HandlerConfig) (context.Context, error) {
+	for _, checker := range rc.RbacCheckers {
+		if err := checker(ctx, api.mustDB(), mux.Vars(req)); err != nil {
```
Comment (fsamin) [inline engine/api/router_middleware_rbac.go:14]: for log purpose, add the name of the checker func as log field

File: engine/api/router_middleware_rbac.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+			return ctx, err
```
Comment (fsamin) [inline engine/api/router_middleware_rbac.go:15]: craft a 403 error

File: engine/api/v2_project.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	"context"
+	"github.com/gorilla/mux"
+	"github.com/ovh/cds/engine/api/rbac"
+	"github.com/rockbears/log"
+	"net/http"
+
+	"github.com/ovh/cds/engine/service"
+)
+
+func (api *API) addVCSOnProjectHandler() (service.Handler, []service.RbacChecker) {
```
Comment (fsamin) [inline engine/api/v2_project.go:13]: it's time to implement a first unit test

File: engine/api/rbac/project.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+	"github.com/go-gorp/gorp"
+	"github.com/rockbears/log"
+
+	"github.com/ovh/cds/engine/api/project"
+	"github.com/ovh/cds/sdk"
+)
+
+func ProjectExist(ctx context.Context, db *gorp.DbMap, vars map[string]string) error {
+	log.Debug(ctx, "Rbac.ProjectExist")
```
Comment (fsamin) [inline engine/api/rbac/project.go:14]: remove this log and add the name of the checker as field of the log (see below)

**Review by fsamin — CHANGES_REQUESTED** [review]

**Review by fsamin — APPROVED** [review]

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17135.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17146.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17148.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by fsamin — CHANGES_REQUESTED** [review]
`+ sed s/Rbac/RBAC/`

**Review by richardlt — CHANGES_REQUESTED** [review]

File: sdk/rbac_project_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	rb := RbacProject{
+		RbacProjectsIDs: []int64{1},
+		All:             false,
+		Role:            RoleCreateProject,
+		RbacGroupsIDs:   []int64{1},
+		RbacUsersIDs:    []string{"aa-aa-aa"},
+	}
+	err := isValidRbacProject("myRule", rb)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "Rbac myRule: role createProject is not allowed on a project permission")
```
Comment (yesnault) [inline sdk/rbac_project_test.go:18]:
```suggestion
	require.Contains(t, err.Error(), fmt.Sprintf("Rbac myRule: role %s is not allowed on a project permission", RoleCreateProject))
```

**Review by yesnault — CHANGES_REQUESTED** [review]

File: engine/api/rbac/dao_rbac.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+		),
+		concat as (
+			SELECT distinct(id) as id FROM (
+				SELECT id FROM userRbac UNION SELECT id FROM groupRbac UNION SELECT id FROM userAllRbac UNION SELECT id FROM groupAllRbac
+			) tmp
+		)
+		SELECT p.* FROM concat c
+		JOIN project p ON p.id = c.id`
+	var projects []sdk.Project
+	if _, err := db.Select(&projects, query, role, userID); err != nil {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/rbac/dao_rbac.go:78]: I have to move this one in project dao

File: engine/api/rbac/rbac.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	for pID := range r.Projects {
+		rbacPrj := &r.Projects[pID]
+		if err := fillRbacProjectWithID(ctx, db, rbacPrj, projectCache, userCache, groupCache); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func fillRbacProjectWithID(ctx context.Context, db gorp.SqlExecutor, rbacPrj *sdk.RbacProject, projectCache map[string]int64, userCache map[string]string, groupCache map[string]int64) error {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/rbac/rbac.go:39]: no, function is test and assert through rbac loader

File: engine/api/rbac/dao_rbac_project.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+		RbacProjectGroupID: groupID,
+	}
+	if err := gorpmapping.InsertAndSign(ctx, db, &rgu); err != nil {
+		return err
+	}
+	return nil
+}
+
+func loadRbacProjectTargeted(ctx context.Context, db gorp.SqlExecutor, rbacProject *rbacProject) error {
+	prjs, err := project.LoadProjectByRbacProject(ctx, db, rbacProject.ID)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/rbac/dao_rbac_project.go:71]: not for project but ok

File: engine/api/group/dao_rbac.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
+	"github.com/ovh/cds/sdk"
+)
+
+func LoadGroupByRbacGlobal(ctx context.Context, db gorp.SqlExecutor, rbacGlobalID int64) ([]sdk.Group, error) {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/group/dao_rbac.go:12]: Need all to check signature

File: engine/api/user/dao_rbac.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	query := `
+		SELECT u.*
+		FROM rbac_global_users rgu
+		JOIN authentified_user u ON u.id = rgu.user_id
+		WHERE rgu.rbac_global_id = $1
+	`
+	return getAll(ctx, db, gorpmapping.NewQuery(query).Args(rbacGlobalID))
+}
+
+func LoadUsersByRbacProject(ctx context.Context, db gorp.SqlExecutor, rbacProjectID int64) ([]sdk.AuthentifiedUser, error) {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/user/dao_rbac.go:22]: Need all to check signature

File: engine/api/v2_project.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+	handler := func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		vars := mux.Vars(r)
+		pKey := vars["projectKey"]
+
+		// my handler
+		log.Info(ctx, "My project: %s", pKey)
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/v2_project.go:19]: will be removed during handler implementation

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17151.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: engine/api/router_middleware_rbac.go [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/ovh/cds/engine/service"
+	"github.com/ovh/cds/sdk"
+	cdslog "github.com/ovh/cds/sdk/log"
+)
+
+func (api *API) rbacMiddleware(ctx context.Context, _ http.ResponseWriter, req *http.Request, rc *service.HandlerConfig) (context.Context, error) {
+	for _, checker := range rc.RbacCheckers {
+		ctx := context.WithValue(ctx, cdslog.RbackCheckerName, sdk.GetFuncName(checker))
+		if err := checker(ctx, api.mustDB(), mux.Vars(req)); err != nil {
+			log.Error(ctx, "%v", err)
```
Comment (fsamin) [inline engine/api/router_middleware_rbac.go:19]: log.ErrorWithStackTrace

**Review by fsamin — CHANGES_REQUESTED** [review]

**Review by yesnault — APPROVED** [review]

**Review by fsamin — APPROVED** [review]

File: engine/api/rbac/dao_rbac_project_group.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+		isValid, err := gorpmapping.CheckSignature(rbacGroups, rbacGroups.Signature)
+		if err != nil {
+			return nil, sdk.WrapError(err, "error when checking signature for rbac_project_groups %d", rbacGroups.ID)
+		}
+		if !isValid {
+			log.Error(ctx, "rbac.getAllRBACProjectGroups> rbac_project_groups %d data corrupted", rbacGroups.ID)
+			continue
+		}
+	}
+	return rbacGroupIDs, nil
```
Comment (richardlt) [inline engine/api/rbac/dao_rbac_project_group.go:60]: Need filter to remove the entity with invalid signature.

File: engine/api/rbac/dao_rbac_project_project.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+		isValid, err := gorpmapping.CheckSignature(projectDatas, projectDatas.Signature)
+		if err != nil {
+			return nil, sdk.WrapError(err, "error when checking signature for rbac_project_projects %d", projectDatas.ID)
+		}
+		if !isValid {
+			log.Error(ctx, "rbac.getAllRBACProjectIdentifiers> rbac_project_projects %d data corrupted", projectDatas.ID)
+			continue
+		}
+	}
+	return rbacProjectIdentifier, nil
```
Comment (richardlt) [inline engine/api/rbac/dao_rbac_project_project.go:29]: Need filter to remove entity with invalid signature.

File: engine/api/rbac/dao_rbac_project_user.go [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+		isValid, err := gorpmapping.CheckSignature(rbacUsers, rbacUsers.Signature)
+		if err != nil {
+			return nil, sdk.WrapError(err, "error when checking signature for rbac_project_users %d", rbacUsers.ID)
+		}
+		if !isValid {
+			log.Error(ctx, "rbac.getAllRBACProjectUsers> rbac_project_users %d data corrupted", rbacUsers.ID)
+			continue
+		}
+	}
+	return rbacProjectUsers, nil
```
Comment (richardlt) [inline engine/api/rbac/dao_rbac_project_user.go:46]: Need filter to remove entities with invalid signature

File: engine/api/rbac/gorp_model.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/ovh/cds/engine/gorpmapper"
+	"github.com/ovh/cds/sdk"
+)
+
+type rbac struct {
+	sdk.RBAC
+	gorpmapper.SignedEntity
+}
+
+func (r rbac) Canonical() gorpmapper.CanonicalForms {
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:14]: Add a line to check that field exists add compilation.

File: engine/api/rbac/gorp_model.go [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+type rbacGlobal struct {
+	ID       int64  `db:"id"`
+	RbacUUID string `db:"rbac_uuid"`
+	sdk.RBACGlobal
+	gorpmapper.SignedEntity
+}
+
+func (rg rbacGlobal) Canonical() gorpmapper.CanonicalForms {
+	return []gorpmapper.CanonicalForm{
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:28]: Add a line to check that fields exists a compilation time.

File: engine/api/rbac/gorp_model.go [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+type rbacGlobalGroup struct {
+	ID                int64 `json:"-" db:"id" yaml:"-"`
+	RbacGlobalID      int64 `db:"rbac_global_id"`
+	RbacGlobalGroupID int64 `db:"group_id"`
+	gorpmapper.SignedEntity
+}
+
+func (rgg rbacGlobalGroup) Canonical() gorpmapper.CanonicalForms {
+	return []gorpmapper.CanonicalForm{
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:54]: Check field exists.

File: engine/api/rbac/gorp_model.go [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+
+type rbacProject struct {
+	ID       int64  `json:"-" db:"id" yaml:"-"`
+	RbacUUID string `json:"-" db:"rbac_uuid" yaml:"-"`
+	sdk.RBACProject
+	gorpmapper.SignedEntity
+}
+
+func (rp rbacProject) Canonical() gorpmapper.CanonicalForms {
+	return []gorpmapper.CanonicalForm{
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:67]: Check field exists

File: engine/api/rbac/gorp_model.go [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+type rbacProjectUser struct {
+	ID                int64  `json:"-" db:"id" yaml:"-"`
+	RbacProjectID     int64  `db:"rbac_project_id"`
+	RbacProjectUserID string `db:"user_id"`
+	gorpmapper.SignedEntity
+}
+
+func (rgu rbacProjectUser) Canonical() gorpmapper.CanonicalForms {
+	return []gorpmapper.CanonicalForm{
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:93]: Check that fields exists

File: engine/api/rbac/gorp_model.go [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+
+type rbacProjectIdentifiers struct {
+	ID            int64 `json:"-" db:"id" yaml:"-"`
+	RbacProjectID int64 `json:"-" db:"rbac_project_id" yaml:"-"`
+	ProjectID     int64 `json:"-" db:"project_id" yaml:"-"`
+	gorpmapper.SignedEntity
+}
+
+func (rpi rbacProjectIdentifiers) Canonical() gorpmapper.CanonicalForms {
+	return []gorpmapper.CanonicalForm{
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:80]: Check that fields exists

File: engine/api/rbac/gorp_model.go [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+
+type rbacProjectGroup struct {
+	ID                 int64 `json:"-" db:"id" yaml:"-"`
+	RbacProjectID      int64 `db:"rbac_project_id"`
+	RbacProjectGroupID int64 `db:"group_id"`
+	gorpmapper.SignedEntity
+}
+
+func (rgg rbacProjectGroup) Canonical() gorpmapper.CanonicalForms {
+	return []gorpmapper.CanonicalForm{
```
Comment (richardlt) [inline engine/api/rbac/gorp_model.go:106]: Check that field exists

**Review by richardlt — CHANGES_REQUESTED** [review]

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17174.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Pull request merged**